File: src/main/java/com/github/stars/service/CloneTaskService.java
package com.github.stars.service;

import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.github.stars.entity.CloneTask;
import com.github.stars.entity.CloneTaskItem;
import com.github.stars.mapper.CloneTaskItemMapper;
import com.github.stars.mapper.CloneTaskMapper;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;

import javax.annotation.Resource;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Clone 任务持久化服务
 */
@Service
public class CloneTaskService {

    @Resource
    private CloneTaskMapper cloneTaskMapper;

    @Resource
    private CloneTaskItemMapper cloneTaskItemMapper;

    /**
     * 创建任务
     */
    public void createTask(CloneTask task) {
        cloneTaskMapper.insert(task);
    }

    /**
     * 获取数据库中最大的 taskCounter 编号，用于服务重启后初始化计数器
     */
    public int getMaxTaskCounterNumber() {
        List<CloneTask> tasks = cloneTaskMapper.selectList(null);
        int maxNum = 0;
        for (CloneTask task : tasks) {
            String taskId = task.getTaskId();
            if (taskId != null && taskId.startsWith("clone_")) {
                try {
                    int num = Integer.parseInt(taskId.substring("clone_".length()));
                    if (num > maxNum) {
                        maxNum = num;
                    }
                } catch (NumberFormatException ignored) {
                    // 忽略非标准格式的 taskId
                }
            }
        }
        return maxNum;
    }

    /**
     * 按主键更新任务
     */
    public void updateTask(CloneTask task) {
        cloneTaskMapper.updateById(task);
    }

    /**
     * 按 taskId 查询单个任务
     */
    public CloneTask getTaskByTaskId(String taskId) {
        LambdaQueryWrapper<CloneTask> wrapper = new LambdaQueryWrapper<>();
        wrapper.eq(CloneTask::getTaskId, taskId);
        return cloneTaskMapper.selectOne(wrapper);
    }

    /**
     * 分页查询所有任务（置顶优先，然后按创建时间降序）
     */
    public Page<CloneTask> getTaskPage(int page, int size) {
        Page<CloneTask> pageParam = new Page<>(page, size);
        LambdaQueryWrapper<CloneTask> wrapper = new LambdaQueryWrapper<>();
        wrapper.orderByDesc(CloneTask::getPinned)
               .orderByDesc(CloneTask::getCreatedAt);
        Page<CloneTask> result = cloneTaskMapper.selectPage(pageParam, wrapper);
        // 对已完成的任务，从 items 表重新计算计数（保证数据一致性）
        for (CloneTask task : result.getRecords()) {
            if (!"RUNNING".equals(task.getStatus()) && !"PENDING".equals(task.getStatus())) {
                recalcFromItems(task);
            }
        }
        return result;
    }

    /** 从 clone_task_item 表重新计算计数 */
    private void recalcFromItems(CloneTask task) {
        int completed = countItemsByTaskIdAndStatus(task.getTaskId(), "CLONED");
        int failed = countItemsByTaskIdAndStatus(task.getTaskId(), "FAILED");
        int skipped = countItemsByTaskIdAndStatus(task.getTaskId(), "SKIPPED");
        if (task.getCompletedRepos() == null || task.getCompletedRepos() != completed
                || task.getFailedRepos() == null || task.getFailedRepos() != failed
                || task.getSkippedRepos() == null || task.getSkippedRepos() != skipped) {
            task.setCompletedRepos(completed);
            task.setFailedRepos(failed);
            task.setSkippedRepos(skipped);
            cloneTaskMapper.updateById(task);
        }
    }

    /**
     * 切换置顶状态
     */
    public boolean togglePin(String taskId) {
        CloneTask task = getTaskByTaskId(taskId);
        if (task == null) return false;
        int newPin = (task.getPinned() != null && task.getPinned() == 1) ? 0 : 1;
        task.setPinned(newPin);
        cloneTaskMapper.updateById(task);
        return newPin == 1;
    }

    /**
     * 分页查询某任务的关联项（按创建时间升序）
     *
     * @param status 可选状态筛选：CLONED/FAILED/SKIPPED，为空时查全部
     */
    public Page<CloneTaskItem> getItemsByTaskId(String taskId, int page, int size, String status) {
        Page<CloneTaskItem> pageParam = new Page<>(page, size);
        LambdaQueryWrapper<CloneTaskItem> wrapper = new LambdaQueryWrapper<>();
        wrapper.eq(CloneTaskItem::getTaskId, taskId);
        if (StringUtils.hasText(status)) {
            wrapper.eq(CloneTaskItem::getStatus, status);
        }
        wrapper.orderByAsc(CloneTaskItem::getCreatedAt);
        return cloneTaskItemMapper.selectPage(pageParam, wrapper);
    }

    /**
     * 插入单项
     */
    public void insertItem(CloneTaskItem item) {
        cloneTaskItemMapper.insert(item);
    }

    /**
     * 删除任务及所有关联项
     */
    public void deleteTaskByTaskId(String taskId) {
        LambdaQueryWrapper<CloneTask> taskWrapper = new LambdaQueryWrapper<>();
        taskWrapper.eq(CloneTask::getTaskId, taskId);
        cloneTaskMapper.delete(taskWrapper);

        LambdaQueryWrapper<CloneTaskItem> itemWrapper = new LambdaQueryWrapper<>();
        itemWrapper.eq(CloneTaskItem::getTaskId, taskId);
        cloneTaskItemMapper.delete(itemWrapper);
    }

    /**
     * 检查是否存在活跃任务（RUNNING 或 PENDING）
     */
    public boolean hasActiveTask() {
        LambdaQueryWrapper<CloneTask> wrapper = new LambdaQueryWrapper<>();
        wrapper.in(CloneTask::getStatus, "RUNNING", "PENDING");
        return cloneTaskMapper.selectCount(wrapper) > 0;
    }

    /**
     * 统计某任务的总项数
     */
    public long countItemsByTaskId(String taskId) {
        LambdaQueryWrapper<CloneTaskItem> wrapper = new LambdaQueryWrapper<>();
        wrapper.eq(CloneTaskItem::getTaskId, taskId);
        return cloneTaskItemMapper.selectCount(wrapper);
    }

    /**
     * 查询所有有失败/跳过项的任务 ID（排除 PENDING）
     */
    public List<String> getTaskIdsWithFailedItems() {
        LambdaQueryWrapper<CloneTaskItem> wrapper = new LambdaQueryWrapper<>();
        wrapper.select(CloneTaskItem::getTaskId)
               .in(CloneTaskItem::getStatus, "FAILED", "SKIPPED")
               .groupBy(CloneTaskItem::getTaskId);
        return cloneTaskItemMapper.selectList(wrapper).stream()
                .map(CloneTaskItem::getTaskId)
                .distinct()
                .collect(Collectors.toList());
    }

    /**
     * 按状态统计任务数
     */
    public long getTaskCountByStatus(String status) {
        LambdaQueryWrapper<CloneTask> wrapper = new LambdaQueryWrapper<>();
        wrapper.eq(CloneTask::getStatus, status);
        return cloneTaskMapper.selectCount(wrapper);
    }

    /**
     * 查询某任务的所有失败项
     */
    public List<CloneTaskItem> getFailedItemsByTaskId(String taskId) {
        LambdaQueryWrapper<CloneTaskItem> wrapper = new LambdaQueryWrapper<>();
        wrapper.eq(CloneTaskItem::getTaskId, taskId)
               .eq(CloneTaskItem::getStatus, "FAILED");
        return cloneTaskItemMapper.selectList(wrapper);
    }

    /**
     * 查询某任务的所有跳过项
     */
    public List<CloneTaskItem> getSkippedItemsByTaskId(String taskId) {
        LambdaQueryWrapper<CloneTaskItem> wrapper = new LambdaQueryWrapper<>();
        wrapper.eq(CloneTaskItem::getTaskId, taskId)
               .eq(CloneTaskItem::getStatus, "SKIPPED");
        return cloneTaskItemMapper.selectList(wrapper);
    }

    /**
     * 更新单项状态
     */
    public void updateItem(CloneTaskItem item) {
        cloneTaskItemMapper.updateById(item);
    }

    /**
     * 按 taskId 和 fullName 查询单项
     */
    public CloneTaskItem getItemByTaskIdAndFullName(String taskId, String fullName) {
        LambdaQueryWrapper<CloneTaskItem> wrapper = new LambdaQueryWrapper<>();
        wrapper.eq(CloneTaskItem::getTaskId, taskId)
               .eq(CloneTaskItem::getFullName, fullName);
        return cloneTaskItemMapper.selectOne(wrapper);
    }

    /**
     * 按 taskId 和 status 统计项数
     */
    public int countItemsByTaskIdAndStatus(String taskId, String status) {
        LambdaQueryWrapper<CloneTaskItem> wrapper = new LambdaQueryWrapper<>();
        wrapper.eq(CloneTaskItem::getTaskId, taskId)
               .eq(CloneTaskItem::getStatus, status);
        return cloneTaskItemMapper.selectCount(wrapper).intValue();
    }
}

File: backend-nestjs/src/clone/services/clone-task.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { PrismaService } from '../../prisma/prisma.service';

@Injectable()
export class CloneTaskService {
    private readonly logger = new Logger(CloneTaskService.name);

    constructor(private readonly prisma: PrismaService) {}

    /**
     * 从数据库恢复最大 taskCounter 编号，确保重启后任务编号连续
     *
     * @returns 当前最大的 taskCounter 数字编号
     */
    async getMaxTaskCounterNumber(): Promise<number> {
        const result = await this.prisma.$queryRawUnsafe<Array<{ num: string }>>(
            `SELECT MAX(CAST(SUBSTRING(task_id, 7) AS UNSIGNED)) AS num FROM clone_task WHERE task_id LIKE 'clone_%'`,
        );
        const num = result?.[0]?.num ? parseInt(result[0].num) : 0;
        return num;
    }

    /**
     * 分页查询克隆任务列表，对已完成/非活跃任务自动修正统计数据
     *
     * @param page 页码
     * @param size 每页数量
     * @returns 任务分页数据
     */
    async getTaskPage(page: number, size: number) {
        const [total, records] = await Promise.all([
            this.prisma.cloneTask.count(),
            this.prisma.cloneTask.findMany({
                orderBy: [{ pinned: 'desc' }, { createdAt: 'desc' }],
                skip: (page - 1) * size,
                take: size,
            }),
        ]);

        // 重新统计已完成任务的计数
        for (const task of records) {
            if (task.status !== 'RUNNING' && task.status !== 'PENDING') {
                const [completed, failed, skipped] = await Promise.all([
                    this.prisma.cloneTaskItem.count({ where: { taskId: task.taskId, status: 'CLONED' } }),
                    this.prisma.cloneTaskItem.count({ where: { taskId: task.taskId, status: 'FAILED' } }),
                    this.prisma.cloneTaskItem.count({ where: { taskId: task.taskId, status: 'SKIPPED' } }),
                ]);
                if (task.completedRepos !== completed || task.failedRepos !== failed || task.skippedRepos !== skipped) {
                    await this.prisma.cloneTask.update({
                        where: { taskId: task.taskId },
                        data: { completedRepos: completed, failedRepos: failed, skippedRepos: skipped },
                    });
                    (task as any).completedRepos = completed;
                    (task as any).failedRepos = failed;
                    (task as any).skippedRepos = skipped;
                }
            }
        }
        return { records, total, size, current: page, pages: Math.ceil(total / size) };
    }

    /**
     * 获取克隆任务详情，包含分页的任务项列表，可按状态筛选
     *
     * @param taskId 任务 ID
     * @param page 页码
     * @param size 每页数量
     * @param status 可选的状态筛选
     * @returns 任务详情及其分页子项，任务不存在时返回 null
     */
    async getTaskDetail(taskId: string, page: number, size: number, status: string) {
        const task = await this.prisma.cloneTask.findUnique({ where: { taskId } });
        if (!task) return null;
        const where: any = { taskId };
        if (status) where.status = status;
        const [total, items] = await Promise.all([
            this.prisma.cloneTaskItem.count({ where }),
            this.prisma.cloneTaskItem.findMany({ where, orderBy: { createdAt: 'asc' }, skip: (page - 1) * size, take: size }),
        ]);
        return { task, items, total, size, current: page, pages: Math.ceil(total / size) };
    }

    /**
     * 分页查询指定任务的任务项列表，可按状态筛选
     *
     * @param taskId 任务 ID
     * @param page 页码
     * @param size 每页数量
     * @param status 可选的状态筛选
     * @returns 任务项分页数据
     */
    async getItemsByTaskId(taskId: string, page: number, size: number, status: string) {
        const where: any = { taskId };
        if (status) where.status = status;
        const [total, records] = await Promise.all([
            this.prisma.cloneTaskItem.count({ where }),
            this.prisma.cloneTaskItem.findMany({ where, orderBy: { createdAt: 'asc' }, skip: (page - 1) * size, take: size }),
        ]);
        return { records, total, size, current: page, pages: Math.ceil(total / size) };
    }

    /**
     * 切换任务的置顶状态
     *
     * @param taskId 任务 ID
     * @returns 是否已置顶
     * @throws 任务不存在时抛出异常
     */
    async togglePin(taskId: string) {
        const task = await this.prisma.cloneTask.findUnique({ where: { taskId } });
        if (!task) throw new Error('任务不存在');
        const newVal = task.pinned === 1 ? 0 : 1;
        await this.prisma.cloneTask.update({ where: { taskId }, data: { pinned: newVal } });
        return newVal === 1;
    }

    /**
     * 根据任务 ID 删除任务及其所有任务项
     *
     * @param taskId 任务 ID
     */
    async deleteTaskByTaskId(taskId: string) {
        await this.prisma.cloneTaskItem.deleteMany({ where: { taskId } });
        await this.prisma.cloneTask.delete({ where: { taskId } });
    }

    /**
     * 检查是否有正在运行或等待中的克隆任务
     *
     * @returns 是否存在活跃任务
     */
    async hasActiveTask() {
        const count = await this.prisma.cloneTask.count({ where: { status: { in: ['RUNNING', 'PENDING'] } } });
        return count > 0;
    }

    /**
     * 查询所有存在失败或跳过项的任务 ID 列表
     *
     * @returns 任务 ID 数组
     */
    async getTaskIdsWithFailedItems(): Promise<string[]> {
        const rows = await this.prisma.cloneTaskItem.findMany({
            where: { status: { in: ['FAILED', 'SKIPPED'] } },
            select: { taskId: true },
            distinct: ['taskId'],
        });
        return rows.map((r) => r.taskId);
    }

    /**
     * 根据任务 ID 查询单个任务
     *
     * @param taskId 任务 ID
     * @returns 任务对象或 null
     */
    async getTaskByTaskId(taskId: string) {
        return this.prisma.cloneTask.findUnique({ where: { taskId } });
    }

    /**
     * 根据任务 ID 和仓库完整名称查询任务项
     *
     * @param taskId 任务 ID
     * @param fullName 仓库完整名称（如 user/repo）
     * @returns 任务项对象或 null
     */
    async getItemByTaskIdAndFullName(taskId: string, fullName: string) {
        return this.prisma.cloneTaskItem.findFirst({ where: { taskId, fullName } });
    }

    /**
     * 统计指定任务中特定状态的任务项数量
     *
     * @param taskId 任务 ID
     * @param status 任务项状态（CLONED/FAILED/SKIPPED 等）
     * @returns 数量
     */
    async countItemsByTaskIdAndStatus(taskId: string, status: string) {
        return this.prisma.cloneTaskItem.count({ where: { taskId, status } });
    }

    /**
     * 统计特定状态的任务总数
     *
     * @param status 任务状态（PENDING/RUNNING/COMPLETED/FAILED）
     * @returns 数量
     */
    async getTaskCountByStatus(status: string) {
        return this.prisma.cloneTask.count({ where: { status } });
    }

    /**
     * 创建一条新的克隆任务项记录
     *
     * @param item 任务项数据，包含 taskId、fullName、status 和可选的 message
     * @returns 创建后的任务项对象
     */
    async insertItem(item: { taskId: string; fullName: string; status: string; message?: string }) {
        return this.prisma.cloneTaskItem.create({ data: { ...item, createdAt: new Date() } });
    }
}
